Specify ntransits

cbrisboi opened the pull request:
This pull request fixes #32.

May not be sufficiently tested (yet), but I believe the essential functionality is there, so I opened the request. 

This allows for subsets of data to be properly analyzed which may only contain (for example) 300 days of data taken over 1500 days. This allows the user to set the HAL plugin to compute the flux over the full 1500 days rather than the 300 days of data contained in the map itself.

cbrisboi committed:
added n_transits to HAL function

cbrisboi committed:
added transit arguments to file reading

cbrisboi committed:
added specifying transits to root analysis bins

cbrisboi committed:
added specifying transits to hdf5 analysis bins

cbrisboi committed:
set all arguments without keywords

cbrisboi committed:
give n_transits to root files

cbrisboi commented:
I guess this causes a few of the tests to fail. I'll take a look this afternoon

henrikef commented:
Good news for you: You can ignore the issues with python 2, and the errors in the python3 tests might be unrelated to your changes. I just opened a [pull request](https://github.com/threeML/hawc_hal/pull/44) to remove the python2 tests and I'm seeing the same errors opening ROOT files 😭

cbrisboi commented:
oh noes, maybe we need to work on the root file reading dependencies next. That is good news for me though I will take another look to make sure. Thanks! 

I have at least one other functional (rather than test related) item I want to check, I want to confirm if LiFF uses sidereal or solar days in its conversion from the hours in the file to ntransits. HAL appears to use 24 hours/transit, and if liff is different I wanted to make that consistent. Its a small difference, but I think consistency would be better than not.

henrikef commented on hawc_hal/HAL.py:
```diff
@@ -47,16 +47,19 @@ class HAL(PluginPrototype):
         self._roi = roi
 
+        # optionally specify n_transits
+        assert (n_transits==None or n_transits > 0.0), "You must specify n_transits >0"
```
Hm. This error message is a bit misleading, as `None` is also a valid value for `n_transits`. Also, do we really need to be that harsh? I think it would be fine to interpret n_transits <= 0 to mean that the user wants to use the transits from the map, similar to what many apps in aerie are doing.

cbrisboi replied:
I agree that using negative numbers might be a bit smoother logically. I can change that.

cbrisboi replied:
This has been fixed in the recent commits

henrikef commented:
Yes, #45 needs to be fixed before this can be merged. Maybe it's an easy fix to the interface, maybe we can fix the version of root to be installed. I'm not sure when I'll be able to work on it though.

The difference between sidereal and solar days is way below our systematic uncertainty, so I wouldn't spend too much time hunting it down at this stage (maybe add a separate issue to come back to later).

Can you please add a simple unit test for this new feature? Doesn't have to be complex, just to show the value of `n_transits` is set correctly in either use case (user provided value or read from file).

henrikef commented:
Any updates on the unit tests for this pull request @cbrisboi ? The issue with the TFile interface has been fixed on the master branch.

maloneka commented:
If I am understanding correctly, this can be merged as soon as the unit test is written. The issue mentioned in #45 was fixed in #48, which has been merged.

henrikef commented:
Yes, I think so. @cbrisboi might have to merge the current master branch into this one to get all the tests to run though. Independently of adding a unit test, it would also be good to see if this actually gives reasonable results on something like the Crab strip.

henrikef commented:
@cbrisboi any updates on this? Will you add some unit tests so we can merge this?

cbrisboi committed:
added maptree ntransits property to allow better testing

cbrisboi committed:
Merge remote-tracking branch 'upstream/master' into specify_ntransits
Need updated hal for gammaln

cbrisboi committed:
made treatment of transits consistent betweeh hdf5 and root files, functions now return transits to pass to Maptree object

cbrisboi committed:
Maptree now take optional n_transits argument, if none given, pulls from data_analysis_bins

cbrisboi committed:
added n_transits test, checks maptree and each data_analysis_bin in it

cbrisboi commented:
@henrikef  I have merged threeML/hawc_hal master branch into this branch and added a unit test. I found during the process of writing the unit test that the hdf5 and root files transit fields were being treated differently, I have fixed that now that it was identified from failing the test.

cbrisboi committed:
unharshed the error message, no longer crashes and gives user info about behavior

henrikef commented on hawc_hal/HAL.py:
```diff
@@ -50,16 +50,25 @@ class HAL(PluginPrototype):
         self._roi = roi
 
+        # optionally specify n_transits
+        if (num_transits > 0.0):
```
shouldn't this be `set_transits` rather than `num_transits`?

cbrisboi replied:
I fixed this in one of my two recent commits

henrikef commented on hawc_hal/maptree/from_hdf5_file.py:
```diff
@@ -66,6 +68,19 @@ def from_hdf5_file(map_tree_file, roi):
+
+    # pick out the transits same as root file
+    use_transits = np.max(transits_bins)
+    if n_transits!=None:
```
This is not mandatory, but for easier readability, I suggest to put the current lines 72 to 76 also into the if block, so that it's easier to see which code is used in what circumstance. Also, there's a much easier way to get the max of some elements in a dataframe. My suggestion:

```
if n_transits is None:
    use_transits = meta_df.iloc[bin_names]["n_transits"].max()
else:
    use_transits = n_transits
```

Or, since `bin_names` always seems to contain all bins (not the user-specified ones), `meta_df["n_transits"].max() ` should work instead and be even faster.

cbrisboi replied:
I went with your second suggestion `meta_df["n_transits"].max()`, as it seemed much cleaner to me.

henrikef reviewed (changes requested):
Hi @cbrisboi , nice work!

Can you please fix the linter error (see my first comment) so that the tests can run?

I'll review again once that's done. We don't expect the tests on conda to pass right now until the next version of threeML gets tagged, but the tests against the master branch of threeML should pass.

cbrisboi committed:
fix Linter error

cbrisboi committed:
access transits array directly in hdf5 files, symmetrize if transit statements for root and hdf5

henrikef reviewed (approved):
Looks good to me now, only thing left is to get the tests passing.

henrikef commented:
Hi @cbrisboi!

Thanks for implementing this!

Please pull in the most recent version of the hawc_hal master into your branch and resolve any conflicts. Note that the `tests` directory moved to `hawc_hal/tests`.

Once the tests pass, this is ready to merge.

henrikef commented:
@cbrisboi has since left the collaboration. Any volunteers to finish this (adding tests) so it can be merged?

henrikef commented:
@xiaojieww Is anyone looking at this?

xiaojieww commented:
> @xiaojieww Is anyone looking at this?

Seems not currently. I plan to solve the #68 first, if still no one working on this I will take a look at it.

torresramiro350 commented:
I think this request can be closed since #84 has been merged.
